src/firmware/app/control/trajectory_planner.c

```diff
@@ -228,155 +128,379 @@ app_trajectory_planner_interpolateConstantPeriodPositionTrajectory(
-                                     2 * arc_segment_length * max_allowable_acceleration);
+        float temp_vel =
+            (float)sqrt(pow(trajectory->trajectory_elements[i - 1].angular_speed, 2) +
+                        2 * segment_lengths[i - 1].angular_segment_length *
+                            max_allowable_acceleration);
 
         // Pick  the lowest of the maximum the available speeds
-        const float lowest_speed =
-            fmin(max_allowable_speed_profile[i], max_allowable_speed);
-        const float speed_to_set = fmin(lowest_speed, temp_vel);
+        const float speed_to_set = fmin(max_allowable_speed, temp_vel);
 
-        velocity_profile[i] = speed_to_set;
+        trajectory->trajectory_elements[i].angular_speed = speed_to_set;
     }
+
+    // The final velocity of an angular profile is assumed zero
+    trajectory->trajectory_elements[num_segments - 1].angular_speed = 0;
 }
 
-void app_trajectory_planner_generateBackwardsContinuousVelocityProfile(
-    const unsigned int num_segments,
-    float forwards_continuous_velocity_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
-    const float arc_segment_length, const float max_allowable_acceleration)
+TrajectoryPlannerGenerationStatus_t
+app_trajectory_planner_modifyTrajectoryToBackwardsContinuous(
+    PositionTrajectory_t* trajectory, TrajectorySegment_t* segment_lengths)
 {
+    const unsigned int num_segments = trajectory->path_parameters.num_segments;
+
+    const float max_allowable_acceleration_linear =
+        trajectory->path_parameters.max_allowable_linear_acceleration;
+    const float max_allowable_acceleration_angular =
+        trajectory->path_parameters.max_allowable_angular_acceleration;
+
     for (unsigned int i = num_segments - 1; i > 0; i--)
     {
+        const float current_speed_linear =
+            trajectory->trajectory_elements[i].linear_speed;
+        const float current_speed_angular =
+            trajectory->trajectory_elements[i].angular_speed;
+        const float previous_speed_linear =
+            trajectory->trajectory_elements[i - 1].linear_speed;
+        const float previous_speed_angular =
+            trajectory->trajectory_elements[i - 1].angular_speed;
+        const float segment_length_linear = segment_lengths[i - 1].linear_segment_length;
+        const float segment_length_angular =
+            segment_lengths[i - 1].angular_segment_length;
+
+
         // Vf = sqrt( Vi^2 + 2*constant_segment_length*max_acceleration)
-        const float temp_vel =
-            (float)sqrt(pow(forwards_continuous_velocity_profile[i], 2) +
-                        2 * arc_segment_length * max_allowable_acceleration);
+        float temp_vel_linear =
+            (float)sqrt(pow(current_speed_linear, 2) +
+                        2 * segment_length_linear * max_allowable_acceleration_linear);
+        float temp_vel_angular =
+            (float)sqrt(pow(current_speed_angular, 2) +
```

Comment:
I did this because sqrt() returns a double and the compiler will give a warning

Reply:
Does your compiler support sqrtf? It should take in a float and return a float instead of doubles.

powf would be a thing in that case as well so you don't need to convert that from double to float before passing into the sqrtf

Reply:
Changed all sqrt() to sqrtf() and pow() to powf()